File: 2022/java/main/Day.java
package main;

import java.util.*;
import java.nio.file.*;
import java.io.File;
import java.io.FileNotFoundException;

public abstract class Day {
    protected final int day;
    protected final List<String> input;

    public Day(int day) {
        this.day = day;
        this.input = this.inputStrings();
    }

    public Object part1() {
        return null;
    }
    
    public Object part2() {
        return null;
    }

    public void printParts() {
        System.out.println(this.part1());
        System.out.println(this.part2());
    }
    
    protected List<String> inputStrings() {
        List<String> data = null;
        String path = String.format("main/data/day%s.txt", this.day);

        try {
            Path p = Paths.get(path);
            data = Files.readAllLines(p);
        }

        catch (Exception e) { e.printStackTrace(); }
        return data;
    }


    protected ArrayList<Integer> inputIntegers() {
        ArrayList<Integer> data = new ArrayList<Integer>();
            
        for (String s : this.input) {
            try { data.add(Integer.parseInt(s)); }
            catch (NumberFormatException e) { }
        }

        return data;
    }


    protected ArrayList<Integer> inputSumBlanks() {
        ArrayList<Integer> data = new ArrayList<Integer>();
        int sum = 0;

        for (String s : this.input) {
            try { sum += Integer.parseInt(s); }

            catch (NumberFormatException e) {
                data.add(sum);
                sum = 0;
            }
        }

        return data;
    }


    protected ArrayList<List<String>> inputStringsBlanks() {
        ArrayList<List<String>> data = new ArrayList<List<String>>();
        int prev = 0;

        for (int i = 0; i < input.size(); i++) {
            String line = input.get(i);
            // Skip unless blank line
            if (!line.trim().isEmpty()) continue;
            
            // Create list from prev until current blank
            data.add(input.subList(prev, i));
            prev = i + 1; 
        }
        
        // Add remaining lines from prev to end
        data.add(input.subList(prev, input.size()));

        return data;
    }


    protected String inputSingleString() {
        return String.join("", input);
    }
}

File: 2022/csharp/Day.cs
using System.Reflection.Metadata.Ecma335;

namespace Application 
{

    abstract class Day
    {
        protected readonly int day;
        protected readonly List<string> data;
        
        public Day(int day) {
            this.day = day;
            this.data = ReadFile();
        }

        public abstract Object Part1();
        public abstract Object Part2();

        public void PrintParts() {
            Console.WriteLine(Part1());
            Console.WriteLine(Part2());
        }

        /*
        Groups lines that are separated by empty lines
        */
        protected List<List<int>> GroupAdjacentNumbers() {
            List<List<int>> result = new();
            List<int> currentGroup = new();

            foreach (string line in data)
            {
                if (string.IsNullOrWhiteSpace(line))
                {
                    result.Add(currentGroup);
                    currentGroup = new List<int>();
                }
                else {
                    currentGroup.Add(int.Parse(line));
                }
            }
            return result;
        }  

        /*
        Returns lines as a single, continous string
        */
        protected string InputSingleString() 
        {
            return string.Join("", data);
        }

         /*
        Reads file into a list of strings
        */
        private List<string> ReadFile() {
            string relativePath = "data/day" + day + ".txt";
            string fullPath = Path.Combine(Environment.CurrentDirectory, relativePath);

            try {
                return new List<string>(File.ReadAllLines(fullPath));
            }
            catch(Exception e) 
            {
                Console.WriteLine(e.Message);
                throw new ApplicationException("Error reading file", e);
            }
        }
    }
}